### PR TITLE
new resource: `azurerm_spring_cloud_elastic_application_performance_monitoring`

### PR DESCRIPTION
**Pull request by ms-henglu**
image
…

### CODE REVIEW

**Review by magodo — CHANGES_REQUESTED** [review]
Thanks for this PR!
I have left some additional mostly minor comments that once addressed this should be good to merge 👍

**Comment by athlete1947** [comment]
Good work guys

File: internal/services/springcloud/spring_cloud_elastic_application_performance_monitoring_resource.go [hunk 1]
```diff
@@ -0,0 +1,305 @@
...
+			Type:         pluginsdk.TypeString,
+			Required:     true,
+			ForceNew:     true,
+			ValidateFunc: validation.StringIsNotEmpty,
+		},
+
+		"spring_cloud_service_id": commonschema.ResourceIDReferenceRequiredForceNew(commonids.SpringCloudServiceId{}),
+
+		"application_packages": {
+			Type:         pluginsdk.TypeString,
```
[the remark this answers is not in the document]
Comment (ms-henglu) [reply internal/services/springcloud/spring_cloud_elastic_application_performance_monitoring_resource.go]: Sure, I've changed it to a list, please take another look, thanks!

**Review by magodo — COMMENTED** [review]
LGTM!

**Comment by ms-henglu** [comment]
Thanks!

File: internal/services/springcloud/spring_cloud_elastic_application_performance_monitoring_resource.go [hunk 2]
```diff
@@ -0,0 +1,310 @@
...
+		Func: func(ctx context.Context, metadata sdk.ResourceMetaData) error {
+			var model SpringCloudElasticApplicationPerformanceMonitoringModel
+			if err := metadata.Decode(&model); err != nil {
+				return fmt.Errorf("decoding: %+v", err)
+			}
+
+			client := metadata.Client.AppPlatform.AppPlatformClient
+			springId, err := commonids.ParseSpringCloudServiceID(model.SpringCloudServiceId)
+			if err != nil {
+				return fmt.Errorf("parsing spring service ID: %+v", err)
```
Comment (stephybun) [inline internal/services/springcloud/spring_cloud_elastic_application_performance_monitoring_resource.go:101]: The parsing functions already return sufficient information in the error messages, there is no need to wrap these. 
```suggestion
				return err
```

File: internal/services/springcloud/spring_cloud_elastic_application_performance_monitoring_resource_test.go [hunk 3]
```diff
@@ -0,0 +1,168 @@
...
+	"github.com/hashicorp/go-azure-sdk/resource-manager/appplatform/2023-09-01-preview/appplatform"
+	"github.com/hashicorp/terraform-provider-azurerm/internal/acceptance"
+	"github.com/hashicorp/terraform-provider-azurerm/internal/acceptance/check"
+	"github.com/hashicorp/terraform-provider-azurerm/internal/clients"
+	"github.com/hashicorp/terraform-provider-azurerm/internal/tf/pluginsdk"
+)
+
+type SpringCloudElasticApplicationPerformanceMonitoringResource struct{}
+
+func TestAccSpringCloudElasticApplicationPerformanceMonitoringPerformanceMonitoring_basic(t *testing.T) {
```
Comment (stephybun) [inline internal/services/springcloud/spring_cloud_elastic_application_performance_monitoring_resource_test.go:19]: All the test names have duplicate `PerformanceMonitoring` at the end
```suggestion
func TestAccSpringCloudElasticApplicationPerformanceMonitoring_basic(t *testing.T) {
```

File: internal/services/springcloud/spring_cloud_elastic_application_performance_monitoring_resource_test.go [hunk 4]
```diff
@@ -0,0 +1,168 @@
...
+			Config: r.basic(data),
+			Check: acceptance.ComposeTestCheckFunc(
+				check.That(data.ResourceName).ExistsInAzure(r),
+			),
+		},
+		data.ImportStep(),
+	})
+}
+
+func TestAccSpringCloudElasticApplicationPerformanceMonitoringPerformanceMonitoring_requiresImport(t *testing.T) {
```
Comment (stephybun) [inline internal/services/springcloud/spring_cloud_elastic_application_performance_monitoring_resource_test.go:33]: same here

File: internal/services/springcloud/spring_cloud_elastic_application_performance_monitoring_resource_test.go [hunk 5]
```diff
@@ -0,0 +1,168 @@
...
+			Config: r.basic(data),
+			Check: acceptance.ComposeTestCheckFunc(
+				check.That(data.ResourceName).ExistsInAzure(r),
+			),
+		},
+		data.RequiresImportErrorStep(r.requiresImport),
+	})
+}
+
+func TestAccSpringCloudElasticApplicationPerformanceMonitoringPerformanceMonitoring_complete(t *testing.T) {
```
Comment (stephybun) [inline internal/services/springcloud/spring_cloud_elastic_application_performance_monitoring_resource_test.go:47]: same here

File: internal/services/springcloud/spring_cloud_elastic_application_performance_monitoring_resource_test.go [hunk 6]
```diff
@@ -0,0 +1,168 @@
...
+			Config: r.complete(data),
+			Check: acceptance.ComposeTestCheckFunc(
+				check.That(data.ResourceName).ExistsInAzure(r),
+			),
+		},
+		data.ImportStep(),
+	})
+}
+
+func TestAccSpringCloudElasticApplicationPerformanceMonitoringPerformanceMonitoring_update(t *testing.T) {
```
Comment (stephybun) [inline internal/services/springcloud/spring_cloud_elastic_application_performance_monitoring_resource_test.go:61]: same here

File: internal/services/springcloud/spring_cloud_elastic_application_performance_monitoring_resource_test.go [hunk 7]
```diff
@@ -0,0 +1,168 @@
...
+		return nil, err
+	}
+	resp, err := client.AppPlatform.AppPlatformClient.ApmsGet(ctx, *id)
+	if err != nil {
+		if response.WasNotFound(resp.HttpResponse) {
+			return pointer.To(false), nil
+		}
+		return nil, fmt.Errorf("retrieving %s: %+v", id, err)
+	}
+	return pointer.To(true), nil
```
Comment (stephybun) [inline internal/services/springcloud/spring_cloud_elastic_application_performance_monitoring_resource_test.go:101]:
```suggestion
	return pointer.To(resp.Model != nil), nil
```

File: internal/services/springcloud/spring_cloud_elastic_application_performance_monitoring_resource_test.go [hunk 8]
```diff
@@ -0,0 +1,168 @@
...
+	template := r.template(data)
+	return fmt.Sprintf(`
+%s
+
+resource "azurerm_spring_cloud_elastic_application_performance_monitoring" "test" {
+  name                    = "acctest-apm-%[2]d"
+  spring_cloud_service_id = azurerm_spring_cloud_service.test.id
+  application_packages    = ["org.example", "org.another.example"]
+  service_name            = "test-service-name"
+  server_url              = "http://127.0.0.1:8200"
```
Comment (stephybun) [inline internal/services/springcloud/spring_cloud_elastic_application_performance_monitoring_resource_test.go:164]: These values are identical to the ones in the `basic` config but they can be updated, can you please change them so we can test the update path for these too?

File: website/docs/r/spring_cloud_elastic_application_performance_monitoring.html.markdown [hunk 9]
```diff
@@ -0,0 +1,83 @@
...
+* `application_packages` - (Required) Specifies a list of the packages which should be used to determine whether a stack trace frame is an in-app frame or a library frame. This is a comma separated list of package names.
+
+* `service_name` - (Required) Specifies the service name which is used to keep all the errors and transactions of your service together and is the primary filter in the Elastic APM user interface.
+
+* `server_url` - (Required) Specifies the server URL. The URL must be fully qualified, including protocol (http or https) and port.
+
+---
+
+* `globally_enabled` - (Optional) Specifies whether the Spring Cloud Application Performance Monitoring resource for Application Insights is enabled globally. Defaults to `false`.
+
```
Comment (stephybun) [inline website/docs/r/spring_cloud_elastic_application_performance_monitoring.html.markdown:61]: We don't separate `Optional` arguments from the required ones
```suggestion
* `globally_enabled` - (Optional) Specifies whether the Spring Cloud Application Performance Monitoring resource for Application Insights is enabled globally. Defaults to `false`.

```

**Review by stephybun — CHANGES_REQUESTED** [review]
Thanks @ms-henglu. I left some comments in-line, once they're fixed up we can take another look through.

**Comment by ms-henglu** [comment]
Hi @stephybun ,

Thanks for the code review, I've updated this PR as suggested. Please take another look, thanks!

**Review by stephybun — APPROVED** [review]
Thanks @ms-henglu and @magodo LGTM 👍

**Pull request merged**